### DISCUSSION TITLE
Request for Support to Download Bibles Through a Proxy Server

### DISCUSSION THREAD

**Discussion by oddpaw** [ezra-bible-app/ezra-bible-app#1093] [Ideas]
The websites for updating modules and downloading Bibles are either inaccessible or subject to censorship for people living in countries that restrict internet access.

Could the program be updated to support setting up a SOCKS5 or HTTP proxy, allowing users to access the internet through a proxy?

If this is too complicated, could you enable support for setting proxies through environment variables like `http_proxy` and `https_proxy`?

Thanks

**Comment by oddpaw** [comment]
One more thing to add: In some countries or within certain companies, access to Dropbox is restricted. Therefore, if proxy support is implemented, it would be great if the program could also use the proxy for synchronization.

**Comment by tobias-klein** [comment]
Hi @oddpaw,

using a proxy for SWORD repository access would require that the SWORD library itself accesses the repositories through a proxy.

Ezra Bible App does not directly interface with the SWORD repositories, but rather it uses the SWORD library to do so. I believe we may need to ask the SWORD library developers whether there is any way to utilize a proxy when connecting to the SWORD repositories. Note that most of the SWORD repositories are still FTP-based. See https://wiki.crosswire.org/Official_and_Affiliated_Module_Repositories

For Dropbox the access to their web-based API is also not implemented directly, but rather through using the Dropbox JavaScript SDK, see https://github.com/dropbox/dropbox-sdk-js.

Therefore, in both cases the setup of the connections is not something we can influence, since it is implemented in libraries. Neither SWORD nor Dropbox currently offer to setup a proxy as part of their library configuration.

Achieving this objective may currently not be possible just based on changing library parameters. Maybe there is a possibility on operating system level, but that is probably more involved ...

Regarding the possibilities with SWORD I'll raise your question on the SWORD Developer's mailinglist to see whether something could be done.

**Comment by tobias-klein** [comment]
Hi @oddpaw!

These are the responses I got on the SWORD mailing list.

### 1) VPN

Has the user tried using a VPN ?
Or are those also banned in the countries?

### 2) proxychains-ng

I'm assuming your SOCKS5 traffic flows through a sufficiently encrypted network tunnel.

For Linux, there are programs which allow to run other programs and direct their network traffic to some SOCKS5 proxy, e.g. proxychains-ng:

  https://github.com/rofl0r/proxychains-ng/

On Debian, Ubuntu and their derivates one can likely install it by using

  sudo apt-get install proxychains4

Proxychains-ng needs to be configured via /etc/proxychains.conf, ~/.proxychains/proxychains.conf or proxychains.conf in the current working directory unless the -f command line option is used to specify a different location. After configuration, one should be able to run programs via commands like

  proxychains4 your_program --with=any arguments

However, the problem with such tools is that they might not always work as intended. For example when network traffic flows via paths which tools like proxychains-ng do not know to intercept. Fpr example, this is sometimes the case for DNS traffic (hostname to IP address lookups) which is sometimes handled via external programs (e.g. DNS cache service on local machine). So be sure to always thorougly test (e.g. using network traffic analysis) whether this actually works properly before actual use, and that nothing leaks. And test again after ANY software updates or configuration changes. So be VERY VERY CAREFUL when using things like proxychains-ng.

A safer option might be to use something like Tails, a Debian Linux based operating system which forces all programs to network over a local SOCKS proxy providing Tor. It might be possible to configure Tails to use some other SOCKS5 proxy as well.

Regarding Tor, please note that in its simplest configuration Tor attempts to connect to public Tor relays, making it possible for eavesdroppers to detect Tor usage. A way around this (as suggested by the Tor project) is to use (private) Tor bridges which use domain fronting, traffic obfuscation and similar tricks. You might also find some of these technologies useful for the tunneling the SOCKS5 traffic.

PS: All security technologies and their implementations, including proxychains-ng, Tails and Tor, have their weak points. So take care when evaluating their fitness for your particular purpose.

**Reply by oddpaw** [reply]
Thank you so much.  
I can use a VPN, although I haven't tested it, but I think the network requests of the Ezra app should be able to go through the VPN.

My operating system is Ubuntu 24.04, and I tested the following methods:

# HTTP proxy environment variables

When I use `http_proxy=http://someip:9999 https_proxy=http://someip:9999 ezra-bible-app` in the command line, the Ezra app does not use these two environment variables. However, when I use `curl`, this curl command can access the network through these proxies, which can be seen in the logs of the proxy client.

# proxychains

When I use `proxychains4 ezra-bible-app` in the command line, the program crashes. It gives a `Trace/breakpoint trap (core dumped)` error. Similarly, `curl` can access the network through proxychains.

# system level Socks5 proxy and HTTP proxy

I set up a socks5 proxy in the system's network settings, and both `curl` and the Chrome browser can access the network through the proxy. However, the Ezra app still doesn't use the system proxy.

# system level FTP proxy

I couldn't find a suitable FTP proxy server software. When I try to set the FTP proxy address to a socks5 proxy or an HTTP proxy, the Ezra app is unable to access the network, resulting in a failure to update the repository.

**Reply by tobias-klein** [reply]
@oddpaw 
Thanks for your feedback.
I believe the least error-prone method is to rely on a VPN.
That is also best from a maintenance perspective.
Testing these SOCKS-related network operations is a lot of effort and it may require debugging on SWORD library level, too. I don't see myself in the position for doing that. I hope you understand.

**Reply by oddpaw** [reply]
@tobias-klein  I totally get it! Thanks a bunch for taking the time to answer my question. I really appreciate you creating such an awesome Bible app!
